h2o-algos/src/main/java/hex/tree/BuildHistogram.java

```diff
@@ -170,102 +203,115 @@ public void setupLocal() {
-        //       Other threads start stealing work from the bottom.
-        //    2) forks the leaf task and (because its polling from the top) executes the LocalMr for the column 0.
-        // This way we should have columns as equally distributed as possible without resorting to shared priority queue
-        new LocalMR(new MrFun() {
-          @Override
-          protected void map(int c) {
-            c = active_cols == null?c:active_cols[c];
-            new LocalMR(new ComputeHistoThread(_hcs.length == 0?new DHistogram[0]:_hcs[c],c,fLargestChunkSz,new AtomicInteger()),numWrks + (c < rem?1:0),ScoreBuildHistogram2.this).fork();
-          }
-        },nactive_cols,ScoreBuildHistogram2.this).fork();
       }
-    }).fork();
+    }
+  }
+  
+  private class PhaseTwo extends H2O.H2OCountedCompleter {
+
+    final int largestChunkSz;
+
+    PhaseTwo(H2O.H2OCountedCompleter cc, int largestChunkSz) {
+      super(cc);
+      this.largestChunkSz = largestChunkSz;
+    }
+    
+    @Override
+    public void onCompletion(CountedCompleter cc){
+      final int ncols = _ncols;
+      final int [] active_cols = _activeCols == null?null:new int[Math.max(1,_activeCols.cardinality())];
+      int nactive_cols = active_cols == null?ncols:active_cols.length;
+      final int numWrks = _hcs.length*nactive_cols < 16*1024?H2O.NUMCPUS:Math.min(H2O.NUMCPUS,Math.max(4*H2O.NUMCPUS/nactive_cols,1));
+      final int rem = H2O.NUMCPUS-numWrks*ncols;
+      BuildHistogram.this.addToPendingCount(1+nactive_cols);
+      if(active_cols != null) {
+        int j = 0;
+        for (int i = 0; i < ncols; ++i)
+          if (_activeCols.contains(i))
+            active_cols[j++] = i;
+      }
+      // MRTask (over columns) launching MrTasks (over number of workers) for each column.
+      // We want FJ to start processing all the columns before parallelizing within column to reduce memory overhead.
+      // (running single column in n threads means n-copies of the histogram)
+      // This is how it works:
+      //    1) Outer MRTask walks down it's tree, forking tasks with exponentially decreasing number of columns until reaching its left most leaf for columns 0.
+      //       At this point, the local fjq for this thread has a task for processing half of columns at the bottom, followed by task for 1/4 of columns and so on.
+      //       Other threads start stealing work from the bottom.
+      //    2) forks the leaf task and (because its polling from the top) executes the LocalMr for the column 0.
+      // This way we should have columns as equally distributed as possible without resorting to shared priority queue
+      new LocalMR(new MrFun() {
+        @Override
+        protected void map(int c) {
+          c = active_cols == null?c:active_cols[c];
+          new LocalMR(new ComputeHistogramTask(_hcs.length == 0?new DHistogram[0]:_hcs[c],c,largestChunkSz,new AtomicInteger()),numWrks + (c < rem?1:0), BuildHistogram.this).fork();
```

Comment:
Could you split this command into more small ones? Or is it memory critical to not save some parts to a variable?  It is very difficult to read.

Reply:
will do